### PR TITLE
Support sanitizer libaries

### PR DESCRIPTION
**Pull request by ligurio**

### CODE REVIEW

File: luzer/tests/test_lib.c [hunk 1]
```diff
@@ -0,0 +1,15 @@
+#include <stdlib.h>
+#include <string.h>
+#include <stdio.h>
+
+#define MESSAGE "Lua"
+
+int
+say_hello(const char *buf) {
```
Comment (Buristan) [inline luzer/tests/test_lib.c:8]: This function is almost the same as in the previous commit. May we declare it non-static and use this library via `ffi.load()` as well?

File: luzer/tests/test_lib.c [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+	if (strncmp(buf, MESSAGE, sizeof(MESSAGE)) == 0) {
+		fprintf(stderr, "Hello, Lua!\n");
+		abort();
+	}
+
+    return 1;
```
Comment (Buristan) [inline luzer/tests/test_lib.c:14]: Nit: mixed tabs and spaces.

**Review by Buristan — APPROVED** [review]
**[PATCH 2/6] luzer: introduce a test with LuaJIT FFI**

Generally, LGTM, with a few nits below.

**Review by Buristan — APPROVED** [review]
**[PATCH 3/6] cmake: refactor SetClangRTLib module**

LGTM.

File: cmake/BuildLibFuzzer.cmake [hunk 3]
```diff
@@ -0,0 +1,77 @@
+# The macro generates a target `build_dso_<name>` that builds
+# a shared library that contains both libFuzzer and a sanitizer.
+#
+# name - a sanitizer name, used in messages and as a postfix in
+#        generated target names.
+# libsanitizer_path - a path to static library with sanitizer.
```
Comment (Buristan) [inline cmake/BuildLibFuzzer.cmake:6]: Nit: IINM, with libFuzzer + sanitizer.

File: cmake/BuildLibFuzzer.cmake [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+)
+
+GEN_BUILD_TARGET("ubsan"
+  ${LIBCLANG_UBSAN_LIB}
+  ${FUZZER_NO_MAIN_LIBRARY}
+  "${LIBCLANG_UBSAN_STRIP}"
+  ${UBSAN_DSO}
+)
+
+GEN_BUILD_TARGET("ubsan_cxx"
```
Comment (Buristan) [inline cmake/BuildLibFuzzer.cmake:72]: Don't we need `asan_cxx` as well? If no add the comment why do we need the separate `ubsan_cxx` target.

File: luzer/tests/CMakeLists.txt [hunk 5]
```diff
@@ -138,8 +138,11 @@ target_link_libraries(luac PRIVATE
 set_target_properties(luac PROPERTIES PREFIX "")
 target_compile_options(luac PRIVATE
   -fsanitize=fuzzer-no-link
+  -fsanitize=undefined
```
Comment (Buristan) [inline luzer/tests/CMakeLists.txt:141]: Why do we need undefined sanitizer in tests?

File: luzer/CMakeLists.txt [hunk 6]
```diff
@@ -1,7 +1,19 @@
...
+else()
+  message(FATAL_ERROR "Unsupported system: ${CMAKE_SYSTEM_NAME}")
+endif()
+
+SetClangLibPath("libclang_rt.asan-${ARCH}.a" LIBCLANG_ASAN_LIB)
+SetClangLibPath("libclang_rt.ubsan_standalone-${ARCH}.a" LIBCLANG_UBSAN_LIB)
+SetClangLibPath("libclang_rt.ubsan_standalone_cxx-${ARCH}.a" LIBCLANG_UBSAN_CXX_LIB)
 SetClangLibPath(${LIBFUZZER_NO_MAIN_LIB_NAME} FUZZER_NO_MAIN_LIBRARY)
 
+include(BuildLibFuzzer)
```
Comment (Buristan) [inline luzer/CMakeLists.txt:15]: Nit: I would rather name it `BuildSanitizers` or whatsoever (since it is more about sanitizers configuration).

File: luzer/tests/CMakeLists.txt [hunk 7]
```diff
@@ -172,7 +177,8 @@ add_test(
 list(APPEND TEST_ENV
   "LD_LIBRARY_PATH=${CMAKE_CURRENT_BINARY_DIR};"
   "LUA_CPATH=${LUA_CPATH};"
-  "LUA_PATH=${LUA_PATH}"
+  "LUA_PATH=${LUA_PATH};"
```
Comment (Buristan) [inline luzer/tests/CMakeLists.txt:180]: This should be the part of the previous commit.

**Review by Buristan — COMMENTED** [review]
[PATCH 4/6] cmake: generate ASan+libFuzzer and UBSan+libFuzzer

Generally LGTM, with a comments and questions below.
Please, add the comment in the commit message _why_ do we need all these workarounds (like strip symbols, etc.).

**Review by Buristan — APPROVED** [review]
**[PATCH 5/6] luzer: rename version.c to config.c**

LGTM.

File: luzer/tests/test_unit.lua [hunk 8]
```diff
@@ -310,4 +310,18 @@ ok, err = pcall(luzer.Fuzz, function() end, nil, 42)
...
+assert(type(luzer.path) == "table")
+
+local so_paths = {
+    "asan",
+    "ubsan",
+    "ubsan_cxx",
+}
+for _, path in ipairs(so_paths) do
+    assert(path)
+    assert(type(path) == "string")
```
Comment (Buristan) [inline luzer/tests/test_unit.lua:324]: Minor: This assert supersedes the previous one.

File: luzer/tests/test_unit.lua [hunk 9]
```diff
@@ -310,4 +310,18 @@ ok, err = pcall(luzer.Fuzz, function() end, nil, 42)
...
+end
```
Comment (Buristan) [inline luzer/tests/test_unit.lua:325]: This checks the content of the constant `so_paths` table, `luzer.path` is not checked here.

**Review by Buristan — COMMENTED** [review]
**[PATCH 6/6] luzer: introduce luzer.path**

Please, consider my comments below.

It would be nice to add tests that contain stack-buffer-overflow (for example) for ASan and UB failed for the UBSan, to be sure that we don't break sanitizers inside the tested library.

File: luzer/tests/luac.c [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+say_hello(lua_State *L) {
+	luaL_checkstring(L, 1);
+	size_t len;
+	const char *buf = lua_tolstring(L, 1, &len);
+	if (strncmp(buf, MESSAGE, sizeof(MESSAGE)) == 0) {
+		fprintf(stderr, "Crash!\n");
+		abort();
+	}
+
+    return 1;
```
[the remark this answers is not in the document]
Comment (ligurio) [reply luzer/tests/luac.c:19]: Fixed:

```diff
diff --git a/luzer/tests/luac.c b/luzer/tests/luac.c
index 7c6c9e0..3883f5f 100644
--- a/luzer/tests/luac.c
+++ b/luzer/tests/luac.c
@@ -16,12 +16,12 @@ say_hello(lua_State *L) {
 		abort();
 	}
 
-    return 1;
+	return 1;
 }
 
 static const struct luaL_Reg functions [] = {
-    { "say_hello", say_hello },
-    { NULL, NULL }
+	{ "say_hello", say_hello },
+	{ NULL, NULL }
 };
 
 int luaopen_luac(lua_State *L) {
@@ -30,5 +30,5 @@ int luaopen_luac(lua_State *L) {
 #else
 	luaL_newlib(L, functions);
 #endif
-    return 1;
+	return 1;
 }
```

File: luzer/tests/luac.c [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+static const struct luaL_Reg functions [] = {
+    { "say_hello", say_hello },
```
[the remark this answers is not in the document]
Comment (ligurio) [reply luzer/tests/luac.c:23]: Fixed:

```diff
diff --git a/luzer/tests/luac.c b/luzer/tests/luac.c
index 7c6c9e0..3883f5f 100644
--- a/luzer/tests/luac.c
+++ b/luzer/tests/luac.c
@@ -16,12 +16,12 @@ say_hello(lua_State *L) {
 		abort();
 	}
 
-    return 1;
+	return 1;
 }
 
 static const struct luaL_Reg functions [] = {
-    { "say_hello", say_hello },
-    { NULL, NULL }
+	{ "say_hello", say_hello },
+	{ NULL, NULL }
 };
 
 int luaopen_luac(lua_State *L) {
@@ -30,5 +30,5 @@ int luaopen_luac(lua_State *L) {
 #else
 	luaL_newlib(L, functions);
 #endif
-    return 1;
+	return 1;
 }
```

File: luzer/tests/luac.c [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+    { NULL, NULL }
+};
+
+int luaopen_luac(lua_State *L) {
+#if LUA_VERSION_NUM == 501
+	luaL_register(L, "luac", functions);
+#else
+	luaL_newlib(L, functions);
+#endif
+    return 1;
```
[the remark this answers is not in the document]
Comment (ligurio) [reply luzer/tests/luac.c:33]: Fixed:

```diff
diff --git a/luzer/tests/luac.c b/luzer/tests/luac.c
index 7c6c9e0..3883f5f 100644
--- a/luzer/tests/luac.c
+++ b/luzer/tests/luac.c
@@ -16,12 +16,12 @@ say_hello(lua_State *L) {
 		abort();
 	}
 
-    return 1;
+	return 1;
 }
 
 static const struct luaL_Reg functions [] = {
-    { "say_hello", say_hello },
-    { NULL, NULL }
+	{ "say_hello", say_hello },
+	{ NULL, NULL }
 };
 
 int luaopen_luac(lua_State *L) {
@@ -30,5 +30,5 @@ int luaopen_luac(lua_State *L) {
 #else
 	luaL_newlib(L, functions);
 #endif
-    return 1;
+	return 1;
 }
```

File: luzer/tests/test_lib.c (hunk 2 above)
Comment (ligurio) [reply luzer/tests/test_lib.c:14]: Fixed:

```diff
@@ -11,5 +11,5 @@ say_hello(const char *buf) {
                abort();
        }
 
-    return 1;
+       return 0;
 }
```

File: luzer/tests/CMakeLists.txt (hunk 5 above)
Comment (ligurio) [reply luzer/tests/CMakeLists.txt:143]: Because test is for symbols required by sanitizer. `luac` is a target for fuzzing built with ubsan, we should check that symbols required by ubsan are provided by our library libfuzzer+sanitizer.

**Comment by ligurio** [comment]
> [PATCH 4/6] cmake: generate ASan+libFuzzer and UBSan+libFuzzer
>
> Generally LGTM, with a comments and questions below.
> Please, add the comment in the commit message why do we need all these workarounds (like strip symbols, etc.).


Updated commit message:

```
commit 3b52503b7e7243200f4afc9cbf00eb97b5b4d8ed
Author: Sergey Bronnikov <estetus@gmail.com>
Date:   Sat Mar 22 11:47:00 2025 +0300

    cmake: generate ASan+libFuzzer and UBSan+libFuzzer
    
    In order for native fuzzing to be effective, native Lua
    extensions must be built with Clang using the argument
    `-fsanitize=fuzzer-no-link` and optionally other sanitizers.
    
    At runtime, you may get an error about missing symbols:
    
    | ./luac.so: undefined symbol: __ubsan_handle_builtin_unreachable
    
    These are libFuzzer symbols, and you must ensure they are
    available in the global namespace. Here are the two options:
    preloading sanitizer and libFuzzer libraries and linking libFuzzer
    into Lua runtime.
    
    The patch implements the first aforementioned option. When luzer
    is built, it attempts to generate custom ASan and UBSan shared
    libraries that have libFuzzer linked in. These files are generated
    using libFuzzer and an appropriate sanitizer library, and preinit
    object files are stripped in static libraries; otherwise, a message
    `.preinit_array section is not allowed in DSO` will prevent
    building DSO. The resulting shared libraries will be called:
    
    - asan_with_fuzzer.so
    - ubsan_with_fuzzer.so
    - ubsan_cxx_with_fuzzer.so
    
    These files must be preloaded to allow use of a sanitizer without
    having to modify Lua runtime. Simply LD_PRELOAD the right .so
    file, and you're good to go.
    
    The idea of the implementation was borrowed from Atheris, see
    details in [1], [2] and [3].
    
    1. https://github.com/google/atheris/commit/6769fecfc3ad41754cbf313f0e4ffe353dca315b
    2. https://github.com/google/atheris/commit/6c17ee407bff81206af2bbd57c480ddd35a449ef
    3. https://github.com/google/atheris/blob/cbf4ad989dcb4d3ef42152990ed89cfceb50e059/native_extension_fuzzing.md
```

File: cmake/BuildLibFuzzer.cmake (hunk 4 above)
Comment (ligurio) [reply cmake/BuildLibFuzzer.cmake]: Honestly, I don't know why Atheris builds only ubsan, asan and ubsan_cxx libraries.

In a private conversation, decided to remove ubsan_cxx flavor and add it later when it will be needed.

Comment (ligurio) [reply cmake/BuildLibFuzzer.cmake]:
```diff
diff --git a/cmake/BuildLibSanitizers.cmake b/cmake/BuildLibSanitizers.cmake
index 6bdeac6..2b58d0a 100644
--- a/cmake/BuildLibSanitizers.cmake
+++ b/cmake/BuildLibSanitizers.cmake
@@ -52,7 +52,6 @@ list(APPEND LIBCLANG_UBSAN_STRIP
 )
 
 set(ASAN_DSO "asan_with_fuzzer.so")
-set(UBSAN_CXX_DSO "ubsan_cxx_with_fuzzer.so")
 set(UBSAN_DSO "ubsan_with_fuzzer.so")
 
 GEN_BUILD_TARGET("asan"
@@ -68,10 +67,3 @@ GEN_BUILD_TARGET("ubsan"
   "${LIBCLANG_UBSAN_STRIP}"
   ${UBSAN_DSO}
 )
-
-GEN_BUILD_TARGET("ubsan_cxx"
-  ${LIBCLANG_UBSAN_CXX_LIB}
-  ${FUZZER_NO_MAIN_LIBRARY}
-  "${LIBCLANG_UBSAN_STRIP}"
-  ${UBSAN_CXX_DSO}
-)
diff --git a/luzer/CMakeLists.txt b/luzer/CMakeLists.txt
index 86f6859..2fb95c3 100644
--- a/luzer/CMakeLists.txt
+++ b/luzer/CMakeLists.txt
@@ -9,7 +9,6 @@ endif()
 
 SetClangLibPath("libclang_rt.asan-${ARCH}.a" LIBCLANG_ASAN_LIB)
 SetClangLibPath("libclang_rt.ubsan_standalone-${ARCH}.a" LIBCLANG_UBSAN_LIB)
-SetClangLibPath("libclang_rt.ubsan_standalone_cxx-${ARCH}.a" LIBCLANG_UBSAN_CXX_LIB)
 SetClangLibPath(${LIBFUZZER_NO_MAIN_LIB_NAME} FUZZER_NO_MAIN_LIBRARY)
 
 include(BuildLibSanitizers)
```

Comment (ligurio) [reply cmake/BuildLibFuzzer.cmake]:
```diff
diff --git a/luzer/config.c b/luzer/config.c
index b1795a6..d7d06dc 100644
--- a/luzer/config.c
+++ b/luzer/config.c
@@ -19,8 +19,3 @@ NO_SANITIZE const char *
 dso_ubsan_string(void) {
 	return "@UBSAN_DSO@";
 }
-
-NO_SANITIZE const char *
-dso_ubsan_cxx_string(void) {
-	return "@UBSAN_CXX_DSO@";
-}
diff --git a/luzer/config.h b/luzer/config.h
index cefb49a..828b588 100644
--- a/luzer/config.h
+++ b/luzer/config.h
@@ -6,6 +6,5 @@ const char *luzer_version_string(void);
 
 const char *dso_asan_string(void);
 const char *dso_ubsan_string(void);
-const char *dso_ubsan_cxx_string(void);
 
 #endif /* LUZER_VERSION_H_ */
diff --git a/luzer/luzer.c b/luzer/luzer.c
index e612e91..8077382 100644
--- a/luzer/luzer.c
+++ b/luzer/luzer.c
@@ -131,7 +131,6 @@ get_coverage_symbols_location(void) {
 
 const char *dso_path_lf_asan;
 const char *dso_path_lf_ubsan;
-const char *dso_path_lf_ubsan_cxx;
 const char *dso_path_libcustom_mutator;
 /* struct paths luzer_paths; */
 
@@ -197,11 +196,6 @@ init(void)
 		perror("access");
 
 	snprintf(path, PATH_MAX, "%s/%s", base_so_path, dso_ubsan_string());
-	dso_path_lf_ubsan_cxx = strdup(path);
-	if (access(path, F_OK))
-		perror("access");
-
-	snprintf(path, PATH_MAX, "%s/%s", base_so_path, dso_ubsan_cxx_string());
 	dso_path_lf_ubsan = strdup(path);
 	if (access(path, F_OK))
 		perror("access");
@@ -359,10 +353,6 @@ luaL_path(lua_State *L) {
 	lua_pushstring(L, dso_path_lf_asan);
 	lua_settable(L, -3);
 
-	lua_pushstring(L, "ubsan_cxx");
-	lua_pushstring(L, dso_path_lf_ubsan_cxx);
-	lua_settable(L, -3);
-
 	lua_pushstring(L, "ubsan");
 	lua_pushstring(L, dso_path_lf_ubsan);
 	lua_settable(L, -3);
diff --git a/luzer/tests/test_unit.lua b/luzer/tests/test_unit.lua
index 3f04133..8bd8517 100644
--- a/luzer/tests/test_unit.lua
+++ b/luzer/tests/test_unit.lua
@@ -317,7 +317,6 @@ assert(type(luzer.path) == "table")
 local so_paths = {
     "asan",
     "ubsan",
-    "ubsan_cxx",
 }
 for _, p in ipairs(so_paths) do
     assert(type(luzer.path[p]) == "string")
```

**Comment by ligurio** [comment]
> [PATCH 6/6] luzer: introduce luzer.path
> Please, consider my comments below.
> It would be nice to add tests that contain stack-buffer-overflow (for example) for ASan and UB failed for the UBSan, to be sure that we don't break sanitizers inside the tested library.

Done:

```diff
diff --git a/luzer/tests/CMakeLists.txt b/luzer/tests/CMakeLists.txt
index 0e44645..311e653 100644
--- a/luzer/tests/CMakeLists.txt
+++ b/luzer/tests/CMakeLists.txt
@@ -138,10 +138,10 @@ target_link_libraries(luac PRIVATE
 set_target_properties(luac PROPERTIES PREFIX "")
 target_compile_options(luac PRIVATE
   -fsanitize=fuzzer-no-link
-  -fsanitize=undefined
+  -fsanitize=address
 )
 
-set(UBSAN_DSO_PATH ${PROJECT_BINARY_DIR}/luzer/${UBSAN_DSO})
+set(ASAN_DSO_PATH ${PROJECT_BINARY_DIR}/luzer/${ASAN_DSO})
 
 add_test(NAME luzer_luac
   COMMAND ${LUA_EXECUTABLE} "${CMAKE_CURRENT_SOURCE_DIR}/test_luac.lua"
@@ -151,10 +151,11 @@ add_test(NAME luzer_luac
 list(APPEND TEST_ENV
   "LUA_CPATH=${LUA_CPATH};"
   "LUA_PATH=${LUA_PATH};"
-  "LD_PRELOAD=${UBSAN_DSO_PATH}"
+  "LD_PRELOAD=${ASAN_DSO_PATH}"
 )
 set_tests_properties(luzer_luac PROPERTIES
   ENVIRONMENT "${TEST_ENV}"
+  PASS_REGULAR_EXPRESSION "ERROR: LeakSanitizer: detected memory leaks"
 )
 add_library(testlib SHARED test_lib.c)
 target_include_directories(testlib PRIVATE ${LUA_INCLUDE_DIR})
diff --git a/luzer/tests/luac.c b/luzer/tests/luac.c
index fa08368..5cfda48 100644
--- a/luzer/tests/luac.c
+++ b/luzer/tests/luac.c
@@ -1,14 +1,25 @@
+#include <assert.h>
+#include <string.h>
+
 #include "lua.h"
 #include "lauxlib.h"
 
-int say_hello(const char *buf);
+int say_hello(const char *buf, size_t len);
 
 static int
 lua_say_hello(lua_State *L) {
 	luaL_checkstring(L, 1);
+	/*
+	 * The length is required, because a string returned by
+	 * lua_tolstring() is always has a zero ('\0') after its last
+	 * character (as in C), but can contain other zeros in its
+	 * body.
+	 */
 	size_t len;
-	const char *buf = lua_tolstring(L, 1, &len);
-	say_hello(buf);
+	const char *str = lua_tolstring(L, 1, &len);
+	/* The length is reduced intentionally. */
+	say_hello(str, len - 1);
+	lua_pop(L, 1);
 	return 0;
 };
 
diff --git a/luzer/tests/test_lib.c b/luzer/tests/test_lib.c
index 20c443e..024c7eb 100644
--- a/luzer/tests/test_lib.c
+++ b/luzer/tests/test_lib.c
@@ -5,8 +5,8 @@
 #define MESSAGE "Hello, Lua!"
 
 int
-say_hello(const char *buf) {
-	if (strncmp(buf, MESSAGE, sizeof(MESSAGE)) == 0) {
+say_hello(const char *buf, size_t len) {
+	if (strncmp(buf, MESSAGE, len) == 0) {
 		fprintf(stderr, "Hello, Lua!\n");
 		abort();
 	}
```

File: docs/api.md [hunk 13]
```diff
@@ -25,6 +25,39 @@ want to allow only those inputs into the corpus that parse successfully. If the
 fuzz target returns `-1` on a given input, `luzer` will not add that input top
 the corpus, regardless of what coverage it triggers.
 
+The `luzer` module provides a function `path()` that returns a table with paths
+to shared libraries with libFuzzer and sanitizers.
+
+```
+> luzer.path
+---
+- ubsan: ./build/luzer/lf_ubsan.so
```
[the remark this answers is not in the document]
Comment (Buristan) [reply docs/api.md]: Last minor note from the offline discussion:
Let's rename library names to the `libluzer_with_{asan,ubsan}.so`?

File: luzer/luzer.c [hunk 14]
```diff
@@ -300,6 +300,10 @@ TestOneInput(const uint8_t* data, size_t size) {
 	lua_State *L = get_global_lua_state();
 
 	char *buf = malloc(size + 1 * sizeof(*buf));
+	if (!buf) {
+		perror("malloc");
+		abort();
```
Comment (Buristan) [inline luzer/luzer.c:305]: I suppose something like `exit()` will be better.

**Review by Buristan — APPROVED** [review]
Hi, Sergey!
Thanks for the fixes!
LGTM, with a [comment](https://github.com/ligurio/luzer/pull/38#discussion_r2182211469) regarding to the previous discussion, and a comment below.

Also, please consider my comment [here](https://github.com/ligurio/luzer/pull/38#discussion_r2182184418).

File: luzer/luzer.c (hunk 14 above)
Comment (ligurio) [reply luzer/luzer.c:305]: Updated:

```diff
--- a/luzer/luzer.c
+++ b/luzer/luzer.c
@@ -310,7 +310,7 @@ TestOneInput(const uint8_t* data, size_t size) {
        char *buf = malloc(size + 1 * sizeof(*buf));
        if (!buf) {
                perror("malloc");
-               abort();
+               exit(EXIT_FAILURE);
        }
        memcpy(buf, data, size);
        buf[size] = '\0';
```

File: docs/api.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (ligurio) [reply docs/api.md]: Updated:

```diff
--- a/cmake/BuildLibSanitizers.cmake
+++ b/cmake/BuildLibSanitizers.cmake
@@ -51,8 +51,8 @@ list(APPEND LIBCLANG_UBSAN_STRIP
   ubsan_init_standalone_preinit.cpp.o
 )
 
-set(ASAN_DSO "asan_with_fuzzer.so")
-set(UBSAN_DSO "ubsan_with_fuzzer.so")
+set(ASAN_DSO "libluzer_with_asan.so")
+set(UBSAN_DSO "libluzer_with_ubsan.so")
 
 GEN_BUILD_TARGET("asan"
   ${LIBCLANG_ASAN_LIB}
```
